### PR TITLE
CppCheck: use some stl algorithms

### PR DESCRIPTION
**Pull request by jhaigh0**
### Description of work

A bit of cppcheck suppression file pruning, in this case dealing with some of the useStlAlgorithm errors.
I got to look at a little under half of these and some I've left in because they'd just make the code much more confusing (which probably means there needs to be some refactoring really).

Hopefully all the changes I've made are for quite simple cases.

I tried to format all the lambda functions nicely but often clang has other ideas.

*There is no associated issue.*

### To test:

- Tests pass
- These do need to be looked over quite carefully, I've tried to keep the functionality completely the same.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Algorithms/src/Bin2DPowderDiffraction.cpp [hunk 1]
```diff
@@ -321,11 +321,12 @@ void Bin2DPowderDiffraction::ReadBinsFromFile(std::vector<double> &Ybins,
...
  */
 size_t Bin2DPowderDiffraction::UnifyXBins(std::vector<std::vector<double>> &Xbins) const {
+  if (Xbins.size() == 0)
+    return 0;
   // get maximal vector size
-  size_t max_size = 0;
-  for (const auto &v : Xbins) {
-    max_size = std::max(v.size(), max_size);
-  }
+  size_t max_size = std::max_element(Xbins.begin(), Xbins.end(), [](const auto a, const auto b) {
```
Comment (robertapplin) [inline Framework/Algorithms/src/Bin2DPowderDiffraction.cpp:327]:
```suggestion
  size_t maxSize = std::max_element(Xbins.cbegin(), Xbins.cend(), [](const auto &a, const auto &b) {
```

File: Framework/CurveFitting/src/Functions/ProcessBackground.cpp [hunk 2]
```diff
@@ -366,11 +366,10 @@ void ProcessBackground::addRegion() {
   }
 
   // Check
-  for (auto it = vx.begin() + 1; it != vx.end(); ++it) {
-    if (*it <= *it - 1) {
-      g_log.error() << "The vector X with value inserted is not ordered incrementally\n";
-      throw std::runtime_error("Build new vector error!");
-    }
+  const auto it = std::adjacent_find(vx.cbegin(), vx.cend(), std::greater_equal<double>());
```
Comment (robertapplin) [inline Framework/CurveFitting/src/Functions/ProcessBackground.cpp:369]: Just want to check this is intended, it uses `std::greater_equal`, but the previous code uses `<=`?

File: Framework/DataHandling/src/SaveISISNexus.cpp [hunk 3]
```diff
@@ -312,10 +312,11 @@ int SaveISISNexus::saveStringVectorOpen(const char *name, const std::vector<std:
     return 0;
   }
   int buff_size = max_str_size;
-  if (buff_size <= 0)
-    for (const auto &str : str_vec) {
-      buff_size = std::max(buff_size, int(str.size()));
-    }
+  if (buff_size <= 0) {
+    buff_size = std::max_element(str_vec.cbegin(), str_vec.cend(), [](const auto a, const auto b) {
```
Comment (robertapplin) [inline Framework/DataHandling/src/SaveISISNexus.cpp:316]:
```suggestion
    buff_size = std::max_element(str_vec.cbegin(), str_vec.cend(), [](const auto &a, const auto &b) {
```

File: Framework/Geometry/src/Instrument.cpp [hunk 4]
```diff
@@ -755,7 +753,7 @@ void Instrument::getBoundingBox(BoundingBox &assemblyBox) const {
       m_cachedBoundingBox = new BoundingBox();
       ComponentID sourceID = getSource()->getComponentID();
       // Loop over the children and define a box large enough for all of them
-      for (auto component : m_children) {
+      for (const auto component : m_children) {
```
Comment (robertapplin) [inline Framework/Geometry/src/Instrument.cpp:756]: Can this be a reference?
```suggestion
      for (const auto &component : m_children) {
```

**Review by robertapplin — CHANGES_REQUESTED** [review]
Code looks good, much easier to read in my opinion! Just had a few minor suggestions, and one question to make sure I don't miss something that may be a difference (but probably not)

File: Framework/CurveFitting/src/Functions/ProcessBackground.cpp (hunk 2 above)
Comment (jhaigh0) [reply Framework/CurveFitting/src/Functions/ProcessBackground.cpp:369]: Yes, this is a bit confusing. In the original, it is checking element `i` is never less than or equal to element `i - 1` (the previous element). So essentially, checking that the vector is strictly (no repeating values) in ordered. You'd think you could just replace this with `std::is_ordered` but that will accept repeating values (`std::is_ordered({1, 1, 2, 3}) == True`).

`std::adjacent_find` searches for a pair `i` and `i + 1` (the next element) which matches a certain condition, in this case `i >= i+1`. If the algorithm finds a pair where this is true, we can also conclude the vector is not strictly ordered, so the two implementations should be equivalent. Maybe it needs a comment.

Comment (robertapplin) [reply Framework/CurveFitting/src/Functions/ProcessBackground.cpp:369]: Thanks for explaining!

File: Framework/Algorithms/src/Bin2DPowderDiffraction.cpp [hunk 5]
```diff
@@ -321,11 +321,12 @@ void Bin2DPowderDiffraction::ReadBinsFromFile(std::vector<double> &Ybins,
...
+  if (Xbins.size() == 0)
+    return 0;
   // get maximal vector size
-  size_t max_size = 0;
-  for (const auto &v : Xbins) {
-    max_size = std::max(v.size(), max_size);
-  }
+  size_t maxSize = std::max_element(Xbins.cbegin(), Xbins.cend(), [](const auto &a, const auto &b) {
+                      return a.size() < b.size();
+                    })->size();
```
Comment (jhaigh0) [inline Framework/Algorithms/src/Bin2DPowderDiffraction.cpp:329]: hopefully fixes clang
```suggestion
                     return a.size() < b.size();
                   })->size();
```

**Review by robertapplin — APPROVED** [review]
Thanks for the changes. Did a code review and I can't see any differences to the old behaviour!

**Pull request merged**